### DISCUSSION TITLE
Micronaut GraphQL Tools

### DISCUSSION THREAD

**Discussion by donbeave** [micronaut-projects/micronaut-core#5541] [Show and tell]
Hello, guys!

I'm using Micronaut for almost a year and quite happy with that experience. One of my API gateway applications is using GraphQL, which relies on `micronaut-graphql` module and [graphql-java-tools](https://github.com/graphql-java-kickstart/graphql-java-tools). The second one is written in Kotlin (which brings Kotlin dependency) and uses reflection quite heavily. After I while when I realized it doesn't work well with Micronaut's annotation interceptors I decided to write my own implementation from scratch with using a similar programming model as GraphQL Java Tools using. Eventually I have an annotation-based implementation which uses Micronauts bean introspections and executable methods processors and leverage zero reflection. I think some other developers are also looking for a similar solution and I'm happy to contribute this code to `micronaut-projects` organization if you guys are thinking it can provide value.

**Comment by graemerocher** [comment]
This sounds awesome. Definitely would be generally useful to all

**Comment by donbeave** [comment]
Hello everyone!

I'm back with the implementation, the source code is located here: https://github.com/donbeave/micronaut-graphql-tools.

There are quite a lot of tests, which almost cover 100% of the code, and also there is one Java example (https://github.com/donbeave/micronaut-graphql-tools/tree/master/doc-examples/example-java), which btw also covered by tests and which can be run as an app to play with GraphQL in the browser by opening http://localhost:8080/graphiql.

I'm still working on writing documentation for this project, I hope I can finish it in two weeks.

I will be quite happy to donate this source code to the Micronaut organization. 

Please let me know what do you think, and what parts you would like to improve.

**Reply by donbeave** [reply]
On top of that I’m being working on Micronaut GraphQL Tools Gradle/Maven codegen plugins. It will automatically generates basic classes, where the developer only needs to fill the business logic, very similar to gRPC development experience.

**Comment by donbeave** [comment]
@graemerocher @sdelamo Sorry for pinging you personally here, I just want to get some feedback to understand how to develop this project later. If you are interested in this implementation, I would like to donate this source code to `micronaut-projects` and support it by adding new features and providing documentation, etc. I can also help with [micronaut-graphql](https://github.com/micronaut-projects/micronaut-graphql) plugin, which looks like a little bit lack of support. From my point of view, the current implementation of Micronaut GraphQL Tools is like a Micronaut alternative to [Spring for GraphQL ](https://spring.io/projects/spring-graphql) project.
If you guys think it can fit the Micronaut ecosystem, I can try to finish it by targeting Micronaut 4.0 release.

**Reply by graemerocher** [reply]
@donbeave sounds awesome! We would love some community investment in GraphQL

**Reply by graemerocher** [reply]
@donbeave do you have some documentation showing how it works?

**Reply by donbeave** [reply]
I'm working on documentation, I will back to this thread again after it will be ready.
